p2p/peer.go

```diff
@@ -348,7 +348,11 @@ func (p *peer) Send(e Envelope) error {
 		// This should never happen.
 		return fmt.Errorf("stream %d not found", e.ChannelID)
 	}
-	err := p.send(e.Message, stream.Write /* blocking */)
+	msgBytes, err := e.marshalMessage()
```

Comment:
Why do we need to marshal msg here? Can't it be done in `send`?

Reply:
`send` function does not have the `Envelope` struct, it only knows `Message`. If we marshal the message in `send` we will marshal X time the same message.
It is why `marshalMessage()` is called here, either it has already been call/done in higher context (Broadcast) and it will return value from the private field `messageBytes` or it will marshal the value.
Envelope struct is required to avoid remarshaling.

Reply:
Can't we refactor `send` to accept `Envelope`?

Reply:
Changes are done 👍